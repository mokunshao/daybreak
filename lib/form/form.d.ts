import React, { FormHTMLAttributes } from 'react';
import './form.scss';
export interface FormValues {
    [K: string]: any;
}
export interface ErrorsMap {
    [k: string]: string;
}
interface Props extends FormHTMLAttributes<HTMLFormElement> {
    values: FormValues;
    fields: Array<{
        name: string;
        label: string;
        input: {
            type: string;
        };
    }>;
    buttons: React.ReactFragment;
    onSubmit: React.FormEventHandler;
    onChange: (values: FormValues) => void;
    errors: {
        [key: string]: string[];
    };
    errorsDisplayMode?: 'first' | 'all';
    transformError?: (erorr: string) => string;
    clearable?: boolean;
}
export declare const Form: React.FunctionComponent<Props>;
export default Form;
//# sourceMappingURL=form.d.ts.map