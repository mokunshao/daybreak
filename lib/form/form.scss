@import '../common.scss';

.daybreak-form {
  &-errors {
    color: $dangerColor;
  }
}
