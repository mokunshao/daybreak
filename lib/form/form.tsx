import React from 'react';
import Input from '../input/input';

export interface FormValues {
  [K: string]: any
}

interface Props {
  values: FormValues
  fields: Array<{ name: string, label: string, input: { type: string } }>
  buttons: React.ReactFragment
  onSubmit: React.FormEventHandler
  onChange: (values: FormValues) => void
  errors: { [key: string]: string[] }
}

const Form: React.FunctionComponent<Props> = (props) => {
  const {
    values,
    fields,
    buttons,
    onSubmit,
    onChange,
    errors,
  } = props;
  const onSubmit2 = (e: React.FormEvent<Element>) => {
    e.preventDefault();
    onSubmit(e);
  };
  const onChange2 = (name: string, value: string) => {
    const data = { ...values, [name]: value };
    onChange(data);
  };
  const onKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.keyCode === 13) {
      e.currentTarget.blur();
    }
  };
  return (
    <form onSubmit={onSubmit2}>
      {fields.map((item) => (
        <div key={item.name}>
          <label htmlFor={item.name}>{item.label}</label>
          <Input
            id={item.name}
            type={item.input.type}
            value={values[item.name]}
            onChange={(e) => onChange2(item.name, e.target.value)}
            onKeyUp={onKeyUp}
          />
          <div>{errors[item.name]}</div>
        </div>
      ))}
      <div>
        {buttons}
      </div>
    </form>
  );
};

export default Form;
