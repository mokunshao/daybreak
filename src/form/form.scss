@import '../common.scss';

.daybreak-form {
  &-errors {
    color: $red;
  }
}
