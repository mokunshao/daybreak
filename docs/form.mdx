---
title: Form
---

import { Playground, Props } from 'docz'
import { useState } from 'react';
import Form from '../src/form/form.tsx';
import { Validator, noError } from '../src/form/validator.tsx';
import { Alert } from '../src/dialog/dialog.tsx';
import { Button } from '../src/button/button.tsx';

# Form

```javascript
import { Form } from 'daybreak';
import { Validator, noError } from 'daybreak';
```

## Validator

Validator is used to validate whether the form data is conforms to the rules.

## noError

`noError` is a helper function to determine whether there is no errors (in other words: whether the data is an empty object). Its effect is similar to `Lodash`'s `isEmpty` function.

## Basic usage

<Playground>
{() => {
  function asyncCheckUserName(username, resolve, reject) {
    setTimeout(() => {
      if (username !== 'admin') {
        resolve();
      } else {
        reject('用户名已经存在!');
      }
    }, 200);
  }
  const customValidator = (username) => (
    new Promise((resolve, reject) => asyncCheckUserName(username, resolve, reject))
  );
  const [formData, setFormData] = useState({
    username: '',
    password: '',
  });
  const [errors, setErrors] = useState({});
  const [fields] = useState([
    { name: 'username', label: '用户名', input: { type: 'text' } },
    { name: 'password', label: '密码', input: { type: 'password' } },
  ]);
  const rules = [
    { key: 'username', require: true },
    { key: 'username', minLength: 3, maxLength: 16 },
    { key: 'username', pattern: /^[A-Za-z0-9]+$/ },
    { key: 'password', require: true },
    { key: 'username', customValidator },
  ];
  const onSubmit = () => {
    Validator(formData, rules, (errorsResult) => {
      setErrors(errorsResult);
      if (noError(errorsResult)) {
        Alert('表单验证通过');
      }
    });
  };
  const onChange = (data) => {
    setFormData(data);
  };
  const buttons = (
    <>
      <Button mode="primary" type="submit">提交</Button>
    </>
  );
  return (
    <div>
      <Form
        values={formData}
        fields={fields}
        buttons={buttons}
        errors={errors}
        onSubmit={onSubmit}
        onChange={onChange}
      />
    </div>
  );
}}
</Playground>

## Validate While Inputing

You may want you validate the form data while inputing, so you can do it with the helpe of `onChange` prop.

<Playground>
{() => {
  function asyncCheckUserName(username, resolve, reject) {
    setTimeout(() => {
      if (username !== 'admin') {
        resolve();
      } else {
        reject('用户名已经存在!');
      }
    }, 200);
  }
  const customValidator = (username) => (
    new Promise((resolve, reject) => asyncCheckUserName(username, resolve, reject))
  );
  const [formData, setFormData] = useState({
    username: '',
    password: '',
  });
  const [errors, setErrors] = useState({});
  const [fields] = useState([
    { name: 'username', label: '用户名', input: { type: 'text' } },
    { name: 'password', label: '密码', input: { type: 'password' } },
  ]);
  const rules = [
    { key: 'username', require: true },
    { key: 'username', minLength: 3, maxLength: 16 },
    { key: 'username', pattern: /^[A-Za-z0-9]+$/ },
    { key: 'password', require: true },
    { key: 'username', customValidator },
  ];
  const onSubmit = () => {
    Validator(formData, rules, (errorsResult) => {
      setErrors(errorsResult);
      if (noError(errorsResult)) {
        Alert('表单验证通过');
      }
    });
  };
  const onChange = (data) => {
    setFormData(data);
    Validator(data, rules, (errorsResult) => {
      setErrors(errorsResult);
    });
  };
  const buttons = (
    <>
      <Button mode="primary" type="submit">提交</Button>
    </>
  );
  return (
    <div>
      <Form
        values={formData}
        fields={fields}
        buttons={buttons}
        errors={errors}
        onSubmit={onSubmit}
        onChange={onChange}
      />
    </div>
  );
}}
</Playground>

## errorsDisplayMode

By default, the validation result will displays only one message per field. If you want to see more than one message, you can set this property to "all".

<Playground>
{() => {
  function asyncCheckUserName(username, resolve, reject) {
    setTimeout(() => {
      if (username !== 'admin') {
        resolve();
      } else {
        reject('用户名已经存在!');
      }
    }, 200);
  }
  const customValidator = (username) => (
    new Promise((resolve, reject) => asyncCheckUserName(username, resolve, reject))
  );
  const [formData, setFormData] = useState({
    username: '',
    password: '',
  });
  const [errors, setErrors] = useState({});
  const [fields] = useState([
    { name: 'username', label: '用户名', input: { type: 'text' } },
    { name: 'password', label: '密码', input: { type: 'password' } },
  ]);
  const rules = [
    { key: 'username', require: true },
    { key: 'username', minLength: 3, maxLength: 16 },
    { key: 'username', pattern: /^[A-Za-z0-9]+$/ },
    { key: 'password', require: true },
    { key: 'username', customValidator },
  ];
  const onSubmit = () => {
    Validator(formData, rules, (errorsResult) => {
      setErrors(errorsResult);
      if (noError(errorsResult)) {
        Alert('表单验证通过');
      }
    });
  };
  const onChange = (data) => {
    setFormData(data);
    Validator(data, rules, (errorsResult) => {
      setErrors(errorsResult);
    });
  };
  const buttons = (
    <>
      <Button mode="primary" type="submit">提交</Button>
    </>
  );
  return (
    <div>
      <Form
        values={formData}
        fields={fields}
        buttons={buttons}
        errors={errors}
        onSubmit={onSubmit}
        onChange={onChange}
        errorsDisplayMode="all"
      />
    </div>
  );
}}
</Playground>

## transformError

By default, validation results are in English, so if you want to display errors in another language, you can set this property.

<Playground>
{() => {
  function asyncCheckUserName(username, resolve, reject) {
    setTimeout(() => {
      if (username !== 'admin') {
        resolve();
      } else {
        reject('用户名已经存在!');
      }
    }, 200);
  }
  const customValidator = (username) => (
    new Promise((resolve, reject) => asyncCheckUserName(username, resolve, reject))
  );
  const [formData, setFormData] = useState({
    username: '',
    password: '',
  });
  const [errors, setErrors] = useState({});
  const [fields] = useState([
    { name: 'username', label: '用户名', input: { type: 'text' } },
    { name: 'password', label: '密码', input: { type: 'password' } },
  ]);
  const rules = [
    { key: 'username', require: true },
    { key: 'username', minLength: 3, maxLength: 16 },
    { key: 'username', pattern: /^[A-Za-z0-9]+$/ },
    { key: 'password', require: true },
    { key: 'username', customValidator },
  ];
  const onSubmit = () => {
    Validator(formData, rules, (errorsResult) => {
      setErrors(errorsResult);
      if (noError(errorsResult)) {
        Alert('表单验证通过');
      }
    });
  };
  const onChange = (data) => {
    setFormData(data);
    Validator(data, rules, (errorsResult) => {
      setErrors(errorsResult);
    });
  };
  const transformError = (error) => {
    const errorsMap = {
      required: '必填',
      minLength: '太短',
      maxLength: '太长',
      pattern: '格式不正确',
    };
    return errorsMap[error];
  };
  const buttons = (
    <>
      <Button mode="primary" type="submit">提交</Button>
    </>
  );
  return (
    <div>
      <Form
        values={formData}
        fields={fields}
        buttons={buttons}
        errors={errors}
        onSubmit={onSubmit}
        onChange={onChange}
        transformError={transformError}
      />
    </div>
  );
}}
</Playground>

## Props

<Props of={Form} />
