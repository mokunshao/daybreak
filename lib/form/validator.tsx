import { FormValues } from './form';

interface FormRule {
  key: string
  require?: boolean
  minLength?: number
  maxLength?: number
  pattern?: RegExp
  validator?: {
    name: string,
    validate: (value: string) => Promise<void>
  }
}

type FormRules = Array<FormRule>;

interface FormErrors {
  [key: string]: string[]
}

function isEmpty(value: any) {
  return value === null || value === undefined || value === '';
}

export function noError(error: object) {
  return Object.keys(error).length === 0;
}

function flat(arr: Array<any>): Array<any> {
  const result: Array<any> = [];
  arr.forEach((item) => {
    if (Array.isArray(item)) {
      result.push(...item);
    } else {
      result.push(item);
    }
  });
  return result;
}

function fromEntries(array: Array<[string, string[]]>): { [key: string]: string[] } {
  const result: { [key: string]: string[] } = {};
  array.forEach(([k, v]) => {
    result[k] = v;
  });
  return result;
}

interface Message {
  text: string
  promise?: Promise<void>
}

const Validator = (values: FormValues, rules: FormRules, callback: Function): void => {
  const errors: any = {};
  const addError = (key: string, message: Message) => {
    if (errors[key] === undefined) {
      errors[key] = [];
    }
    errors[key].push(message);
  };
  rules.forEach((rule) => {
    const { key } = rule;
    const value = values[key];
    if (rule.validator) {
      const promise = rule.validator.validate(value);
      addError(key, { text: '用户名已经存在', promise });
    }
    if (rule.require && isEmpty(value)) {
      addError(key, { text: '必填' });
    }
    if (rule.minLength && !isEmpty(value) && value.length < rule.minLength) {
      addError(key, { text: '太短' });
    }
    if (rule.maxLength && !isEmpty(value) && value.length > rule.maxLength) {
      addError(key, { text: '太长' });
    }
    if (rule.pattern && !(rule.pattern.test(value))) {
      addError(key, { text: '格式不合法' });
    }
  });

  const PromiseList = flat(Object.values(errors)).filter((item) => item.promise).map((item) => (
    item.promise
  ));

  Promise.all(PromiseList).then(() => {
    const newErrors = fromEntries(
      Object.keys(errors).map((key) => [key, errors[key].map((item: Message) => item.text)]),
    );
    callback(newErrors);
  }, () => {
    const newErrors = fromEntries(
      Object.keys(errors).map((key) => [key, errors[key].map((item: Message) => item.text)]),
    );
    callback(newErrors);
  });
};

export default Validator;
export { Validator };
