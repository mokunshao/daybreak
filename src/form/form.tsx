import React, { Fragment, FormHTMLAttributes } from 'react';
import Input from '../input/input';
import { classes } from '../utils/classes';
import { joinedClass } from '../utils/joinedClass';
import './form.scss';

const form = joinedClass('form');

export interface FormValues {
  [K: string]: any;
}

export interface ErrorsMap {
  [k: string]: string;
}

interface Props extends FormHTMLAttributes<HTMLFormElement> {
  values: FormValues;
  fields: Array<{ name: string; label: string; input: { type: string; }; }>;
  buttons: React.ReactFragment;
  onSubmit: React.FormEventHandler;
  onChange: (values: FormValues) => void;
  errors: { [key: string]: string[]; };
  errorsDisplayMode?: 'first' | 'all';
  transformError?: (erorr: string) => string;
}

const Form: React.FunctionComponent<Props> = (props) => {
  const {
    values,
    fields,
    buttons,
    onSubmit,
    onChange,
    errors,
    className,
    errorsDisplayMode = 'first',
    transformError,
  } = props;
  const onSubmit2 = (e: React.FormEvent<Element>) => {
    e.preventDefault();
    onSubmit(e);
  };
  const onChange2 = (name: string, value: string) => {
    const data = { ...values, [name]: value };
    onChange(data);
  };
  const onKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.keyCode === 13) {
      e.currentTarget.blur();
    }
  };
  const transformError2 = (error: string) => {
    const errorsMap: ErrorsMap = {
      required: 'required',
      minLength: 'too short',
      maxLength: 'too long',
      pattern: 'wrong format',
    };
    return (transformError && transformError(error)) || errorsMap[error] || error;
  };
  const showErrors = (name: string) => {
    if (errors[name]) {
      switch (errorsDisplayMode) {
        case 'first':
          return transformError2(errors[name][0]);
        case 'all':
          return errors[name].map(transformError2).join(', ');
        default:
          return transformError2(errors[name][0]);
      }
    }
    // return <>&nbsp;</>;
    return null;
  };
  return (
    <form onSubmit={onSubmit2} className={classes(form(), className)}>
      <table>
        <tbody>
          {fields.map((item) => (
            <Fragment key={item.name}>
              <tr>
                <td>
                  <label htmlFor={item.name}>{item.label}</label>
                </td>
                <td>
                  <Input
                    id={item.name}
                    type={item.input.type}
                    value={values[item.name]}
                    onChange={(e) => onChange2(item.name, e.target.value)}
                    onKeyUp={onKeyUp}
                  />
                </td>
              </tr>
              <tr>
                <td />
                <td>
                  <small className={form('errors')}>
                    {showErrors(item.name)}
                  </small>
                </td>
              </tr>
            </Fragment>
          ))}
        </tbody>
        <tfoot>
          <tr>
            <td />
            <td>
              {buttons}
            </td>
          </tr>
        </tfoot>
      </table>
    </form>
  );
};

export default Form;
