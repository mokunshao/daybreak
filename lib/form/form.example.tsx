import React, { useState } from 'react';
import Form, { FormValues } from './form';
import { Validator } from './validator';

export default () => {
  const [formData, setFormData] = useState<FormValues>({
    username: 'test',
    password: 'test',
  });
  const [fields] = useState([
    { name: 'username', label: '用户名', input: { type: 'text' } },
    { name: 'password', label: '密码', input: { type: 'password' } },
  ]);
  const onSubmit = () => {
    const rules = [
      { key: 'username', require: true },
      { key: 'username', minLength: 3, maxLength: 16 },
      { key: 'username', pattern: /^[A-Za-z0-9]+$/ },
      { key: 'password', require: true },
    ];
    const errors = Validator(formData, rules);
    console.log(errors);
  };
  const onChange = (data: FormValues) => setFormData(data);
  return (
    <div>
      <Form
        values={formData}
        fields={fields}
        buttons={
          (
            <>
              <button type="submit">提交</button>
              <button type="button">返回</button>
            </>
          )
        }
        onSubmit={onSubmit}
        onChange={onChange}
      />
    </div>
  );
};
