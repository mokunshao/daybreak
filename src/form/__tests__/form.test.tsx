import { render } from '@testing-library/react';
import React, { useState } from 'react';
import { Form } from '../form';
import { Validator, noError } from '../validator';

describe('Form', () => {
  it('exit', () => {
    expect(Form).toBeTruthy();
  });
  it('match snapshot', () => {
    function Example() {
      const [formData, setFormData] = useState({
        username: '',
        password: '',
      });
      const errors = {};
      const [fields] = useState([
        { name: 'username', label: '用户名', input: { type: 'text' } },
        { name: 'password', label: '密码', input: { type: 'password' } },
      ]);
      const onSubmit = () => {
      };
      const onChange = (data: any) => {
        setFormData(data);
      };
      const buttons = (
        <>
          <button type="submit">提交</button>
        </>
      );
      return (
        <div>
          <Form
            values={formData}
            fields={fields}
            buttons={buttons}
            errors={errors}
            onSubmit={onSubmit}
            onChange={onChange}
          />
        </div>
      );
    }
    const { asFragment } = render(<Example />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('validator', () => {
  it('validator', () => {
    const formData = {
      username: '',
      password: '',
    };
    const rules = [
      { key: 'username', require: true },
      { key: 'username', minLength: 3, maxLength: 16 },
      { key: 'username', pattern: /^[A-Za-z0-9]+$/ },
      { key: 'password', require: true },
    ];
    Validator(formData, rules, (errorsResult: any) => {
      expect(errorsResult).toStrictEqual({ username: ['required', 'pattern'], password: ['required'] });
    });
  });
  it('noError', () => {
    expect(noError({})).toBe(true);
    expect(noError({ username: ['required'] })).toBe(false);
  });
});
