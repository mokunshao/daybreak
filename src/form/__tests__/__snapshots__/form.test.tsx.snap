// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form match snapshot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="daybreak-form"
    >
      <table>
        <tbody>
          <tr>
            <td>
              <label
                for="username0.7081323259541124"
                style="display: block;"
              >
                用户名
              </label>
            </td>
            <td>
              <div
                class="daybreak-input-wrapper"
              >
                <input
                  class="daybreak-input"
                  id="username0.7081323259541124"
                  type="text"
                  value=""
                />
                <div
                  class="daybreak-input-icons"
                />
              </div>
            </td>
          </tr>
          <tr>
            <td />
            <td>
              <small
                class="daybreak-form-errors"
              />
            </td>
          </tr>
          <tr>
            <td>
              <label
                for="password0.7081323259541124"
                style="display: block;"
              >
                密码
              </label>
            </td>
            <td>
              <div
                class="daybreak-input-wrapper"
              >
                <input
                  class="daybreak-input"
                  id="password0.7081323259541124"
                  type="password"
                  value=""
                />
                <div
                  class="daybreak-input-icons"
                >
                  <span
                    class="daybreak-input-icon"
                    title="show password"
                  >
                    <svg
                      class="daybreak-icon daybreak-input-icon-password"
                    >
                      <use
                        xlink:href="#show"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </td>
          </tr>
          <tr>
            <td />
            <td>
              <small
                class="daybreak-form-errors"
              />
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <td />
            <td>
              <button
                type="submit"
              >
                提交
              </button>
            </td>
          </tr>
        </tfoot>
      </table>
    </form>
  </div>
</DocumentFragment>
`;
